dwst/scripts/particles.js

```diff
@@ -20,7 +20,32 @@ const specialChars = [
   '\\',
 ];
 
-export class InvalidParticles extends Error { }
+export class InvalidParticles extends Error {
+  constructor(expression, expected, expectedStringLength, remainder) {
+    super();
+    this.expression = expression;
+    this.expected = expected;
+    // this.got = got;
+    this.remainder = remainder;
+    this.expectedStringLength = expectedStringLength;
+  }
+  getGot() {
+    return this.remainder.substring(0, this.expectedStringLength);
+  }
+  getErrorPosition() {
+    return this.expression.length - this.remainder.length + this.expectedStringLength - 1;
+  }
+}
+
+export function invalidToMlog(invalidParticle) {
```

Comment:
I think this should be moved to a new `errors.js` file. Eventually it would be good to have all error messages in the same file. Perhaps we wish to do this over time. Maybe we could start with this one? It should probably be called `invalidParticlesError` to match the name of the exception.

Reply:
done